containers/src/Data/Tree.hs

```diff
@@ -181,14 +182,55 @@ mfixTree f
 instance Traversable Tree where
     traverse f (Node x ts) = liftA2 Node (f x) (traverse (traverse f) ts)
 
+-- | Folds in preorder
+
+-- See Note [Implemented Foldable Tree functions]
 instance Foldable Tree where
-    foldMap f (Node x ts) = f x `mappend` foldMap (foldMap f) ts
+    fold = foldMap id
+    {-# INLINABLE fold #-}
+
+    foldMap f = go
+      where go (Node x ts) = f x `mappend` foldMap go ts
+    {-# INLINE foldMap #-}
+
+    foldr f z = \t -> go t z  -- Use a lambda to allow inlining with two arguments
```

Comment:
Right, I don't think it would be bad to do that, but all `foldrs` in containers are defined as `foldr f z` (even when z can be eta-reduced, not sure why), so I've kept it consistent.
I also noticed this brought up in [GHC !5259](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/5259) for `foldl'`. But it seems that it was not definitively addressed, and in the end they went with `foldl' f z0`.

Reply:
OK. There might be/have been a reason. The secrets of the GHC inliner....